src/chips/sam4l/flashcalw.rs

```diff
@@ -0,0 +1,1103 @@
+#[derive(Clone, Copy, PartialEq)]
+pub enum FlashCMD {
+    NOP,
+    WP,
+    EP,
+    CPB,
+    LP,
+    UP,
+    EA,
+    WGPB,
+    EGPB,
+    SSB,
+    PGPFB,
+    EAGPF,
+    QPR,
+    WUP,
+    EUP,
+    QPRUP,
+    HSEN,
+    HSDIS,
+}
+
+//The two Flash speeds
+#[derive(Clone, Copy)]
+pub enum Speed {
+    Standard,
+    HighSpeed
+}
+
+/// FlashState is used to track the current state of the flash in high level
+/// command.
+#[derive(Clone, Copy, PartialEq)]
+pub enum FlashState {
+    Locking,
+    Unlocking,
+    WritePageBuffer,
+    Writing,
+    Reading,
+    Erasing,
+    Ready,
+    Unconfigured
+}
+
+// The FLASHCALW controller
+pub struct FLASHCALW {
+    registers: *mut Registers,
+    ahb_clock: pm::Clock,
+    hramc1_clock: pm::Clock,
+    pb_clock: pm::Clock,
+    wait_until_ready: fn(&FLASHCALW) -> (),
+    error_status: TakeCell<u32>,
+    ready: TakeCell<bool>,
+    client: TakeCell<&'static Client>,
+    current_state: Cell<FlashState>,
+    current_command: Cell<Command>,
+    page: Cell<i32>,
+    page_buffer: TakeCell<[u8; FLASH_PAGE_SIZE as usize]>
+}
+
+//static instance for the board. Only one FLASHCALW on chip.
+pub static mut flash_controller : FLASHCALW = 
+    FLASHCALW::new(FLASHCALW_BASE_ADDRS, pm::HSBClock::FLASHCALW, 
+        pm::HSBClock::FLASHCALWP, pm::PBBClock::FLASHCALW);
+
+
+// Few constants relating to module configuration.
+const FLASH_PAGE_SIZE : u32 = 512;
+const FLASH_NB_OF_REGIONS : u32 = 16;
+const FLASHCALW_REGIONS : u32 = FLASH_NB_OF_REGIONS;
+const FLASHCALW_CMD_KEY : u32 = 0xA5 << 24;
+
+#[cfg(CONFIG_FLASH_READ_MODE_HIGH_SPEED_DISABLE)]
+const FLASH_FREQ_PS1_FWS_1_FWU_MAX_FREQ : u32 = 12000000;
+#[cfg(CONFIG_FLASH_READ_MODE_HIGH_SPEED_DISABLE)]
+const FLASH_FREQ_PS0_FWS_0_MAX_FREQ : u32 = 18000000;
+#[cfg(CONFIG_FLASH_READ_MODE_HIGH_SPEED_DISABLE)]
+const FLASH_FREQ_PS0_FWS_1_MAX_FREQ : u32 = 36000000;
+#[cfg(CONFIG_FLASH_READ_MODE_HIGH_SPEED_DISABLE)]
+const FLASH_FREQ_PS1_FWS_0_MAX_FREQ : u32 = 8000000;
+
+
+#[cfg(not(CONFIG_FLASH_READ_MODE_HIGH_SPEED_DISABLE))]
+const FLASH_FREQ_PS2_FWS_0_MAX_FREQ : u32 = 24000000;
+
+const GP_ALL_FUSES_ONE : u64 = !0 as u64;
+
+// Macros for getting the i-th bit.
+macro_rules! get_bit {
+    ($w:expr) => (0x1u32 << $w);
+}
+
+/// This is simply std::cmp::min from std
+#[inline]
+fn min<T: Ord>(v1: T, v2: T) -> T {
+    if v1 <= v2 { v1 } else { v2 }    
+}
+
+// Should never ever have to wait (because we're using interrupts!)
+pub fn default_wait_until_ready(flash : &FLASHCALW) {
+    while !flash.get_ready_status() { 
+        panic!("Flash wasn't ready for a command -- interrupt system used incorrectly.");
+    }
+}
+
+/// Trait for a client of the flash controller.
+pub trait Client {
+    //  Called upon a completed call
+    fn command_complete(&self, err: Error);     
+}
+    
+impl FLASHCALW {
+
+    pub fn mark_ready(&self) {
+        self.ready.put(Some(true));
+    }
+
+
+    pub fn get_ready_status(&self) -> bool {
```

Comment:
This method is _probably_ not necessary, but if it is, it's _waaaay_ over complicated. Changing `self.ready` to a `Cell` fixes this I believe.